paderbox/array/kernel.py

```diff
@@ -0,0 +1,212 @@
+import typing
+import numpy as np
+import paderbox as pb
+
+
+__all__ = [
+    'np_kernel',
+    'ai_dilate',
+    'ai_erode',
+    'max_kernel',
+    'min_kernel',
+    'mean_kernel',
+    'median_kernel',
+]
+
+
+def _plot(a):
+    """
+    >>> _plot([0, 1, 2, 3, 4, 5, 6, 7, 8])
+    ' ▁▂▃▄▅▆▇█'
+    """
+    a = np.array(a)
+    values = np.array(list(' ▁▂▃▄▅▆▇█'))
+    m = np.amax(a)
+    if m == 0:
+        m = 1
+    a = np.round(a * ((len(values)-1) / m)).astype(int)
+    print(repr(''.join(values[a])))
+
+
+def np_kernel1d(
+        x,
+        kernel_size,  # an odd number
+        *,
+        kernel: callable,  # e.g. np.mean, np.amax, np.amin
+        padding_mode: 'typing.Literal["edge"]' = 'edge',
+        pad_position: 'typing.Literal["pre", "post", None]' = 'pre',
+):
+    """
+    Apply a kernel to the last axis of x.
+
+    Args:
+        x: np.array
+        kernel_size:
+        kernel: a function that should be applied to the kernel window.
+            Has to accept an axis argument, like np.mean, np.amax, np.median.
+        padding_mode:
+            The mode for np.pad, see np.pad for the options.
+        pad_position:
+            'pre': First pad and then apply the kernel.
+            'post': First apply the kernel then pad.
+            None: Don't pad and return an array with a smaller last axis.
+
+    Returns:
+
+    >>> a = np.array([0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0,])
+    >>> _plot(a); _plot(np_kernel1d(a, 3, kernel=np.amax))
+    '    ████    '
+    '   ██████   '
+    >>> _plot(a); _plot(np_kernel1d(a, 3, kernel=np.amin))
+    '    ████    '
+    '     ██     '
+    >>> _plot(a); _plot(np_kernel1d(a, 3, kernel=np.mean))
+    '    ████    '
+    '   ▃▅██▅▃   '
+
+    """
+    assert kernel_size % 2 == 1, (kernel_size, 'kernel size has to be odd.')
```

Comment:
Is it better to follow Python style guidelines and use `ValueError` for parameter checks? `asserts` can be deactivated by the user and carry less information

Reply:
The way you formulate it, the "assert" is correct here.
You let the program run once and see, if it doesn't fail.
Then you disable all asserts and get rid of the overhead for the checks.

Reply:
This is a function that is used outside of `paderbox`. In that case, `ValueError` is preferred. But that's also a matter of taste

Reply:
I don't care about the Exception type, when the exception should never be caught.
Hyperparameters are such a case for me. I find asserts are easier to read and more compact.
It is ok to deactivate `asserts`. The idea is to reduce the overhead, and that is the case here.